### PR TITLE
Add vctrs support to lubridate

### LINKED ISSUE #721: vctrs integration
Just a placeholder for now, but once vctrs is on lubridate, it would be worth implementing `vec_type2()`, `vec_cast()`, and `vec_arith()` methods here. This could potentially eliminate the need for S4, but that may require too much work to be worthwhile. 

I've included some code I used for testing vctrs below, it's likely to need updates, but should be a place to start. I'll be happy to help out when the time comes.

```R
#' @examples
#' w1 <- lubridate::years(1)
#' w2 <- lubridate::ddays(7)
#' w3 <- lubridate::interval("2020-01-01", "2020-01-08")
#'
#' vec_ptype(w1)
#' vec_ptype(w2)
#' vec_ptype(w3)
#'
#' vec_c(w1, w3, .ptype = w3)
#' vec_c(w1, w2)
#' vec_c(w2, w3)
#'
#' library(lubridate)
#' vec_c(years(1), months(1), weeks(1), days(1))
#' vec_c(dyears(1), years(1))
NULL

# https://github.com/tidyverse/lubridate/issues/707
new_Period <- function() lubridate::seconds(integer())
new_Duration <- function() lubridate::dseconds(integer())
new_Interval <- function() lubridate::interval(character(), character())

# vec_type2 ---------------------------------------------------------------

vec_type2.Period <- function(x, y) UseMethod("vec_type2.Period")
vec_type2.Duration <- function(x, y) UseMethod("vec_type2.Duration")
vec_type2.Interval <- function(x, y) UseMethod("vec_type2.Interval")

vec_type2.Period.NULL <- function(x, y) x[0L]
vec_type2.Duration.NULL <- function(x, y) x[0L]
vec_type2.Interval.NULL <- function(x, y) x[0L]

vec_type2.Period.default <- function(x, y) stop_incompatible_type(x, y)
vec_type2.Duration.default <- function(x, y) stop_incompatible_type(x, y)
vec_type2.Interval.default <- function(x, y) stop_incompatible_type(x, y)

vec_type2.Period.Period     <- function(x, y) new_Period()

vec_type2.Period.Duration   <- function(x, y) new_Period()
vec_type2.Duration.Period   <- function(x, y) new_Period()

vec_type2.difftime.Period   <- function(x, y) new_Period()
vec_type2.Period.difftime   <- function(x, y) new_Period()

vec_type2.Period.Interval   <- function(x, y) new_Period()
vec_type2.Interval.Period   <- function(x, y) new_Period()

vec_type2.Duration.Duration <- function(x, y) new_Duration()

vec_type2.difftime.Duration <- function(x, y) new_Duration()
vec_type2.Duration.difftime <- function(x, y) new_Duration()

vec_type2.Duration.Interval <- function(x, y) new_Duration()
vec_type2.Interval.Duration <- function(x, y) new_Duration()

# vec_type2.difftime.difftime <- function(x, y) new_Duration()
vec_type2.difftime.Interval <- function(x, y) new_difftime()
vec_type2.Interval.difftime <- function(x, y) new_difftime()

vec_type2.Interval.Interval <- function(x, y) new_Interval()


# vec_cast ----------------------------------------------------------------

vec_cast.Period <- function(x, to) UseMethod("vec_cast.Period")
vec_cast.Duration <- function(x, to) UseMethod("vec_cast.Duration")
vec_cast.Interval <- function(x, to) UseMethod("vec_cast.Interval")

vec_cast.Period.NULL <- function(x, to) x
vec_cast.Duration.NULL <- function(x, to) x
vec_cast.Interval.NULL <- function(x, to) x

vec_cast.Period.default <- function(x, to) stop_incompatible_cast(x, to)
vec_cast.Duration.default <- function(x, to) stop_incompatible_cast(x, to)
vec_cast.Interval.default <- function(x, to) stop_incompatible_cast(x, to)

vec_cast.Period.Period     <- function(x, to) x
vec_cast.Period.Duration   <- function(x, to) lubridate::as.period(x)
vec_cast.Period.difftime   <- function(x, to) lubridate::as.period(x)
vec_cast.Period.Interval   <- function(x, to) lubridate::as.period(x)

vec_cast.Duration.Duration <- function(x, to) x
vec_cast.Duration.Period   <- function(x, to) lubridate::as.duration(x)
vec_cast.Duration.difftime <- function(x, to) lubridate::as.duration(x)
vec_cast.Duration.Interval <- function(x, to) lubridate::as.duration(x)

vec_cast.Interval.Interval <- function(x, to) x
```

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #721 
Fixes @batpigandme's comment in https://github.com/r-lib/vctrs/issues/919#issuecomment-600680606

This PR adds vctrs support for Period / Duration / Interval S4 classes. Specifically, each one now has methods for `vec_ptype2()`, `vec_cast()`, `vec_proxy()`, `vec_restore()`, `vec_proxy_compare()`, and `vec_proxy_equal()`.

I have added an extensive test suite, testing the cast/ptype behavior, and also the behavior of various vctrs functions (ordering, binding, c, etc). I think these vctrs functionality tests are good because this is one way to ensure that your `vec_restore()` and `vec_proxy()` methods are right.

- There is no common type between Period / Duration / Interval. This could be added later if required, but for now this seems safe and should solve 95% of the problems.

- Duration and difftime have a common type.

- I have not prevented inheritance here (like we did in glue), as we expect inheritance in the cast/ptype2 methods to be prevented natively in vctrs in the near future.

- Period and Interval types use a data frame as their proxy, with a special optional `rcrd_names` column to hold any vector names. This is better than adding row names, because it allows duplicates. The comparison/equality proxies do not have this column.

- Interval types have a time zone attached. I followed the same rules as POSIXct when determining the common type.

   - Local time zones are overruled by the other time zone
   - Otherwise the time zone used in the common type comes from the LHS

- You'll see a `TODO` in the constructor section. Lubridate constructors are rather slow, and I'd really like to store a global empty `Period` and `Duration` object, but I can't seem to create one in the `.onLoad()`. If I call `period()` in `.onLoad()`, for some reason the S4 `initialize()` method gets bypassed, but the validity method still gets run and we get an error from there. I'm curious if anyone has ever seen this.

- This currently requires dev vctrs, but it is in Suggests.

- There was also a small bug where `interval(tzone = "my-tz")` wasn't respecting the `tzone` argument. This has been fixed.

### CODE REVIEW

**Comment by DavisVaughan** [comment]
Travis and Windows failures seem unrelated to me (Travis looks like the Rcpp problem)

File: R/intervals.r [hunk 1]
```diff
@@ -205,7 +205,10 @@ interval <- function(start = NULL, end = NULL, tzone = tz(start)) {
   }
 
   if (length(start) == 0 || length(end) == 0) {
-    start <- POSIXct()
+    if (missing(tzone)) {
```
Comment (vspinu) [inline R/intervals.r]: `missing` is a bad programming practice + this logic is not correct. We want to force tzone and fail if it's called on wrong type.

**Comment by hadley** [comment]
We'd really like to get this into the next release as it makes lubridate classes work reliably in tidyr and dplyr 1.0.0.

**Comment by vspinu** [comment]
Ok, let's push it with the next release.

**Comment by DavisVaughan** [comment]
It does currently require dev vctrs because of a bug that I fixed with `vec_chop()`, which affects `vec_rbind()` (at the very least). It caused a crash of R.

Maybe we could do one more small vctrs release with that fix before the big one with the breaking changes?

File: R/intervals.r (hunk 1 above)
Comment (DavisVaughan) [reply R/intervals.r]: Okay, I think this is definitely a bug in some way though. If a `tzone` is specified but no `start` is specified, then an empty `UTC` POSIXct is used as the start, and the `tzone` is out of sync.

Open to suggestions!

``` r
library(lubridate)
#> 
#> Attaching package: 'lubridate'
#> The following object is masked from 'package:base':
#> 
#>     date

str(interval(tzone = "America/New_York"))
#> Formal class 'Interval' [package "lubridate"] with 3 slots
#>   ..@ .Data: num(0) 
#>   ..@ start: 'POSIXct' num(0) 
#>  - attr(*, "tzone")= chr "UTC"
#>   ..@ tzone: chr "America/New_York"
```

<sup>Created on 2020-03-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Comment (vspinu) [reply R/intervals.r]: Good catch. The logic there is a bit convoluted. I will have a look tomorrow.

File: R/vctrs.R [hunk 2]
```diff
@@ -0,0 +1,370 @@
+#' vctrs methods for lubridate
+#'
+#' @keywords internal
+#' @name lubridate-vctrs
+NULL
+
+# ------------------------------------------------------------------------------
+# Constructors
+
+# TODO: Find a way to generate these once at load time for efficiency. S4
```
Comment (hadley) [inline R/vctrs.R]: What about with `delayedAssign()`? (we know it's not triggered by package building)

**Comment by vspinu** [comment]
vctrs tests are failing with CRAN vctrs, so I am afraid this will have to go in next release in a month or two. Will pull into dev as soon as new version of vctrs is on CRAN as well.

**Comment by DavisVaughan** [comment]
Noting that it would be nice to add `vec_arith()` methods as well. I was just bitten by this, and was very confused.

```r
library(vctrs)
library(lubridate)

i <- structure(c(0, 1, 2, 3), class = "Date")
i
#> [1] "1970-01-01" "1970-01-02" "1970-01-03" "1970-01-04"

i - lubridate::days(1)
#> [1] "1969-12-31" "1970-01-01" "1970-01-02" "1970-01-03"

# This "works" because `days()` is a Period, which is an S4 declared to be a
# "double".
# The `.Data` slot of Period is the number of seconds, which is 0 here.
# So `i` is returned unchanged.
vec_arith("-", i, lubridate::days(1))
#> [1] "1970-01-01" "1970-01-02" "1970-01-03" "1970-01-04"
```

**Comment by vspinu** [comment]
@DavisVaughan What is the status of `vctrs`? I would like to push a bug fix release to CRAN towards the end of the week. Is this one ready to be included?

**Comment by hadley** [comment]
Would it be possible to hold off release until the end of next week? vctrs is scheduled for release on  2020-05-06, so @DavisVaughan could update this PR with changes needed for the latest version of vctrs, and we could get this PR merged before dplyr 1.0.0 is out.

**Comment by vspinu** [comment]
Sure, next week then!

**Comment by vspinu** [comment]
@DavisVaughan could you please update on the status? Thanks!

**Comment by DavisVaughan** [comment]
@vspinu just bumped to use CRAN vctrs. It should be good to merge.

`vec_arith()` isn't exactly right in vctrs yet so we are waiting to add support for it in lubridate. It shouldn't have any effects on how lubridate is used with dplyr though.

**Pull request merged**